File: 08-10-2025_Successful_Pairs_of_Spells_and_Potions.java
import java.util.Arrays;

class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        int n = spells.length;
        int m = potions.length;
        int[] res = new int[n];

        for (int i = 0; i < n; i++) {
            int left = 0, right = m - 1;
            int ans = m; 

            while (left <= right) {
                int mid = left + (right - left) / 2;
                long product = (long) spells[i] * potions[mid];

                if (product >= success) {
                    ans = mid;
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }

            res[i] = m - ans;
        }

        return res;
    }
}

File: 08-10-2025_Successful_Pairs_of_Spells_and_Potions.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {

        int n = spells.size();

        sort(potions.begin(),potions.end());
        
        vector<int> res(n,0);
    
        for(int i=0;i<n;i++)
        {
            int left = 0;
            int right = potions.size()-1;

            int m = potions.size();
            int ans = m;
            
            while(left<=right)
            {   
                
                int mid = left + (right-left)/2;

                long long product = static_cast<long long>(spells[i]) *potions[mid]  ;

                if(product>=success)
                {
                    ans = mid;
                    right = mid -1;
                }

                else
                {
                    left = mid+1;
                }
            }


            res[i] = m - ans;
        }


        return res;
        
    }
};
